sdk/src/Core/Amazon.Util/AWSSDKUtils.cs

```diff
@@ -444,6 +422,51 @@ public static IEnumerable<string> SplitResourcePathIntoSegments(string resourceP
             return resolvedSegments;
         }
 
+        /// <summary>
+        /// This is used to split resource path into segments for composing a URL, not for signing. When splitting resource path into 
+        /// segments for composing a URL we must discern that which is a literal which we do not encode and that which is a label which
+        /// we do encode.
+        /// </summary>
+        /// <param name="resourcePath">The resource path, which includes literals and labels.</param>
+        /// <param name="pathResources">The dictionary for substituting label values.</param>
+        /// <returns></returns>
+        public static IEnumerable<UriComponent> SplitResourcePathIntoSegmentsV2(string resourcePath, IDictionary<string, string> pathResources)
+        {
+            var splitChars = new char[] { SlashChar };
+            var pathSegments = resourcePath.Split(splitChars, StringSplitOptions.None);
```

Comment:
With IndexOf and slicing the resource path as you go you can avoid the additional string allocations which makes this code way more efficient. This would also avoid the additional char array allocation. See https://github.com/aws/aws-sdk-net/pull/3365/files for an example

Reply:
I did some digging and found that using the built in Split method requires far less complexity especially for the corner cases (see below for more details). For that reason I decided to stick with the `Split` method. I did a quick benchmark with the following code (first part is a snippet)
```
            while (true)
            {
                var slashIndex = resourcePathSpan.IndexOf(SlashChar);
                if (slashIndex == -1)
                {
                    break;
                }

                ProcessSegment(resourcePathSpan.Slice(0, slashIndex).ToString(), uriComponentSegments);
                resourcePathSpan = resourcePathSpan.Slice(slashIndex + 1);
            }

            // Handle the remaining part of the resourcePathSpan
            if (!resourcePathSpan.IsEmpty)
            {
                ProcessSegment(resourcePathSpan.ToString(), uriComponentSegments);
            }

        private void ProcessSegment(string segment, List<UriComponent> uriComponentSegments)
        {
            if (!pathResources.TryGetValue(segment, out var pathSegment))
            {
                uriComponentSegments.Add(new UriComponent { SegmentType = SegmentType.Literal, Value = segment });
            }
            else if (segment.EndsWith("+}", StringComparison.Ordinal))
            {
                var resolvedSegment = pathSegment.AsSpan();
                while (true)
                {
                    var innerSlashIndex = resolvedSegment.IndexOf(SlashChar);
                    if (innerSlashIndex == -1)
                    {
                        break;
                    }
                    string temp = resolvedSegment.Slice(0, innerSlashIndex).ToString();
                    uriComponentSegments.Add(new UriComponent { Value = temp, SegmentType = SegmentType.Label });
                    resolvedSegment = resolvedSegment.Slice(innerSlashIndex + 1);
                }
            }
            else
            {
                uriComponentSegments.Add(new UriComponent { SegmentType = SegmentType.Label, Value = pathSegment });
            }
        }
```

For example, in the case above if there is a greedy path and it is the last portion of the string, then simply checking if the `slashIndex != -1` won't cut it because it's possible that the resource path doesn't end in a slash. This then requires us to either create a new version of the resourcePath which ends in a `/` which allocates more memory since we have to make a copy of the string or to check again after the loop exits to see if the string is empty. But even that may not be enough because within the greedy path logic we also check for a `/` which may not exist at the end.


The happy path, however, did result in a decrease in allocations. 

```
| Method           | Mean     | Error   | StdDev  | Median   | Gen0   | Allocated |
|----------------- |---------:|--------:|--------:|---------:|-------:|----------:|
| TestRegularSplit | 197.9 ns | 3.90 ns | 9.93 ns | 194.5 ns | 0.0503 |     632 B |
| TestSpanAndSlice | 172.8 ns | 3.42 ns | 5.42 ns | 170.4 ns | 0.0393 |     496 B |
```